tests/integration/targets/odbc/tasks/main.yml

```
@@ -0,0 +1,146 @@
+- include_tasks: no_pyodbc.yml
+  when: ansible_os_family != 'FreeBSD' and ansible_os_family != 'Suse' and ansible_os_family != 'Debian'
+
+#
+# Get pyodbc installed
+#
+- include_tasks: install_pyodbc.yml
+
+#
+# Test missing parameters & invalid DSN
+#
+- include_tasks: negative_tests.yml
+
+#
+# Setup DSN per env
+#
+- name: Changing DSN for Suse
+  set_fact:
+    dsn: "DRIVER={PSQL};Server=localhost;Port=5432;Database=postgres;Uid={{ my_user }};Pwd={{ my_pass_decrypted }};UseUnicode=True"
+  when: ansible_os_family == 'Suse'
+
+- name: Changing DSN for Debian
+  set_fact:
+    dsn: "DRIVER={PostgreSQL Unicode};Server=localhost;Port=5432;Database=postgres;Uid={{ my_user }};Pwd={{ my_pass_decrypted }};UseUnicode=True"
+  when: ansible_os_family == 'Debian'
+
+#
+# Name setup database
+#
+- name: Create a user to run the tests with
+  postgresql_user:
+    name: "{{ my_user }}"
+    password: "{{ my_pass }}"
+    encrypted: 'yes'
+    role_attr_flags: "SUPERUSER"
+    db: postgres
+  become_user: "{{ pg_user }}"
+  become: True
+
+- name: Create a table
+  odbc:
+    dsn: "{{ dsn }}"
+    query: |
+        CREATE TABLE films (
+            code        char(5) CONSTRAINT firstkey PRIMARY KEY,
+            title       varchar(40) NOT NULL,
+            did         integer NOT NULL,
+            date_prod   date,
+            kind        varchar(10),
+            len         interval hour to minute
+        );
+  become_user: "{{ pg_user }}"
+  become: True
+  register: results
+
+- debug:
+    var: results
+
+- assert:
+    that:
+      - results is changed
+
+#
+# Insert records
+#
+- name: Insert a record without params
+  odbc:
+    dsn: "{{ dsn }}"
+    query: "INSERT INTO films (code, title, did, date_prod, kind, len) VALUES ('asdfg', 'My First Movie', 1, '2019-01-12', 'SyFi', '02:00')"
+  become_user: "{{ pg_user }}"
+  become: True
+  register: results
+
+- assert:
+    that:
+      - results is changed
+
+- name: Insert a record with params
+  odbc:
+    dsn: "{{ dsn }}"
+    query: "INSERT INTO films (code, title, did, date_prod, kind, len) VALUES (?, ?, ?, ?, ?, ?)"
+    params:
+      - 'qwert'
+      - 'My Second Movie'
+      - 2
+      - '2019-01-12'
+      - 'Comedy'
+      - '01:30'
+  become_user: "{{ pg_user }}"
+  become: True
+  register: results
+
+- assert:
+    that:
+      - results is changed
+
+#
+# Select data
+#
+- name: Perform select single row without params
+  odbc:
+    dsn: "{{ dsn }}"
+    query: "SELECT * FROM films WHERE code='asdfg'"
+  register: results
+
+- debug:
+    var: results
+
+- assert:
+    that:
+      - results is successful
+      - results.row_count == 1
+
+- name: Perform select multiple rows with params
+  odbc:
+    dsn: "{{ dsn }}"
+    query: 'SELECT * FROM films WHERE code=? or code=?'
+    params:
+      - 'asdfg'
+      - 'qwert'
+  register: results
+
+- debug:
+    var: results
+
+- assert:
+    that:
+      - results is successful
```

```suggestion
    that:
      - results is not changed ?
      - results is successful
```
interesting what result.changed returns when SELECT

---

Again with this being a generic ODBC connector I always set `changed = True` like the command module. I see in the mysql_query its inspecting the SQL statement and trying to determine if its a DDL/DML query but with ODBC we would have to consider every back end. For example, some DBs support the `SELECT INTO` syntax. Or I could see someone having a strange use case where an app has its own derivative of SQL(i.e. SWQL [solar winds query language]) that supports ODBC connections (http://www.solarwinds.com/documentation/en/flarehelp/sam/content/sam-odbc-user-experience-monitor-sw3272.htm) but may have its own keywords that would be DDL/DML. Like the command module I think the best we could do with ODBC is to have the users leverage the changed_when function. We could add something like that to the documentation/examples. What are your thoughts on this?

---

For example, in postgres we could catch something like `INSERT 0 1` when 1 row has been inserted or `DELETE 0` when nothing has been deleted. So we can catch something and handle (though i don't remember how it's implemented in postgresql or mysql_query modules).
Doesn't ODBC provide anything so that we can understand there were any changes or not?
I've never used this driver.

If you feel any difficulties let's complete the module as it is now but we should have `- result is changed` everywhere in the tests (because it's expected behavior now) and a note in the `note:` doc section that the module always returns 'changed: true` even if there are no changes (like when SELECT or DELETE 0).